src/components/CsvUpload/index.tsx

```diff
@@ -0,0 +1,60 @@
+import { PlusOutlined } from "@ant-design/icons";
+import { Flex } from "antd";
+import { RcFile } from "antd/es/upload";
+import Dragger from "antd/es/upload/Dragger";
+import { connect } from "react-redux";
+import React, { ReactElement } from "react";
+
+import CsvRequest, { DEFAULT_CSV_DATASET_KEY } from "../../state/image-dataset/csv-dataset";
+import { ImageDataset, Megaset } from "../../state/image-dataset/types";
+import metadataStateBranch from "../../state/metadata";
+import {
+    ReceiveAvailableDatasetsAction,
+    ReceiveImageDatasetAction,
+} from "../../state/metadata/types";
+import selectionStateBranch from "../../state/selection";
+import { ChangeSelectionAction } from "../../state/selection/types";
+
+type CsvUploadProps = {
+    replaceImageDataset: (dataset: ImageDataset) => ReceiveImageDatasetAction;
+    receiveAvailableDatasets: (megasets: Megaset[]) => ReceiveAvailableDatasetsAction;
+    changeDataset: (id: string) => ChangeSelectionAction;
+};
+
+/**
+ * An upload area for CSV data. When CSV data is provided, replaces the current image dataset
+ * with a new `CsvRequest` image dataset and triggers the loading of the CSV data.
+ */
+function CsvUpload(props: CsvUploadProps): ReactElement {
+    const action = async (file: RcFile): Promise<string> => {
+        // TODO: handle loading via URL
+        const fileContents = await file.text();
+        const dataset = new CsvRequest(fileContents);
+        props.replaceImageDataset(dataset);
+
+        // CSV Request mocks up a single dataset
+        const megasets = await dataset.getAvailableDatasets();
+        props.receiveAvailableDatasets(megasets);
+        props.changeDataset(DEFAULT_CSV_DATASET_KEY);
+
+        return Promise.resolve("");
```

Comment:
this should all be in a logic, which may help you avoid connecting this component directly to state in the first place

Reply:
this is the only outstanding issue I have now